### PR TITLE
Lint Python code with ruff

### PR DESCRIPTION
**Pull request by cclauss**
### Short description

Lint Python code with [`ruff`](https://docs.astral.sh/ruff) which is a superset of `black`, `isort`, `mccabe`, `pylint`, `pylint-django`, and numerous other Python linting tools.  It is written in Rust to be able to consistently lint and format the repo in well under one second.

### Proposed changes

- Lint Python code with ruff.
- Make minor fixes to enable several linters.
- Put `ruff format` in place but disable it until a follow-on pull request.

### Resolved issues

Fixes: #2442

__________________________________________________

[Pull Request Review Guidelines](https://digitalfabrik.github.io/integreat-cms/pull-request-review-guide.html)

### CODE REVIEW

File: integreat_cms/core/signals/auth_signals.py [hunk 1]
```diff
@@ -48,7 +48,7 @@ def user_logged_in_callback(
 # pylint: disable=unused-argument
 @receiver(user_logged_out)
 def user_logged_out_callback(
-    sender: ModelBase, request: HttpRequest | HttpRequest, user: User, **kwargs: Any
+    sender: ModelBase, request: HttpRequest, user: User, **kwargs: Any
```
Comment (cclauss) [inline integreat_cms/core/signals/auth_signals.py:51]: Was this second one meant to have been [`django.http.HttpResponse`](https://docs.djangoproject.com/en/stable/ref/request-response/#httpresponse-objects)?

File: integreat_cms/cms/models/events/recurrence_rule.py [hunk 2]
```diff
@@ -167,7 +167,7 @@ def advance() -> Iterator[date]:
         i = 0
         end_date = self.recurrence_end_date or date.max
         while True:
-            for _ in advance():
+            for __ in advance():
```
Comment (timobrembeck) [inline integreat_cms/cms/models/events/recurrence_rule.py:170]:
```suggestion
            for _recurrence in advance():
```
Just a personal preference, but even if the variable is unused, I'd like to indicate its purpose.

File: integreat_cms/core/signals/auth_signals.py (hunk 1 above)
Comment (timobrembeck) [reply integreat_cms/core/signals/auth_signals.py:51]: No, I think this was a remnant of some more fine-graded typing like `WSGIRequest | ASGIRequest` which I overlooked during a global search/replace...
So your suggestion looks good!

File: .circleci/config.yml [hunk 3]
```diff
@@ -85,6 +85,20 @@ jobs:
...
+      - image: cimg/python:3.11.7
+    resource_class: small
+    steps:
+      - checkout
+      - attach_workspace:
+          at: .
+      - run:
+          name: Run ruff
+          command: |
+            pip install --user ruff
```
Comment (timobrembeck) [inline .circleci/config.yml:99]:
```suggestion
```
This should not be required when the ruff job depends on `pip-install`.

**Review by timobrembeck — CHANGES_REQUESTED** [review]
Great, thanks a lot! :+1:
I really like ruff, and the potential consolidation of tools.
This means we can now remove isort & black, right? Pylint does not seem to be completely implemented yet (https://github.com/astral-sh/ruff/issues/970)?
This is a cool start to enforce the rules, I'm fine with ignoring a lot of them at the moment and gradually implement these changes later down the road.

File: pyproject.toml [hunk 4]
```diff
@@ -287,6 +287,7 @@ dev-pinned = [
     "requests-toolbelt==1.0.0",
     "rfc3986==2.0.0",
     "rich==13.7.0",
+    "ruff==0.3.0",
```
[the remark this answers is not in the document]
Comment (cclauss) [reply pyproject.toml:290]: Running `./tools/update_dependencies.sh` would require that I downgrade both `node` and `python3` which will break other projects that I need to be working.

`ruff` is written in Rust so it has no Python sub-dependencies....
% `python3 -m venv .venv`
% `source .venv/bin/activate`
% `pip install ruff`
```
Collecting ruff
  Using cached ruff-0.3.0-py3-none-macosx_10_12_x86_64.macosx_11_0_arm64.macosx_10_12_universal2.whl.metadata (23 kB)
Using cached ruff-0.3.0-py3-none-macosx_10_12_x86_64.macosx_11_0_arm64.macosx_10_12_universal2.whl (14.9 MB)
Installing collected packages: ruff
Successfully installed ruff-0.3.0
```

[the remark this answers is not in the document]
Comment (cclauss) [reply pyproject.toml:290]: Perhaps if you could create and merge the `./tools/update_dependencies.sh` pull request then I could rebase this.

[the remark this answers is not in the document]
Comment (timobrembeck) [reply pyproject.toml:290]: Ah great, then the only thing we need to preserve ruff for future executions of the script is adding it to the non-pinned dev extra:
https://github.com/digitalfabrik/integreat-cms/blob/41b995e1304285c2d6853813a3168b672af840db/pyproject.toml#L80

[the remark this answers is not in the document]
Comment (cclauss) [reply pyproject.toml]: Great.  That is _much_ easier.  Sorry that I did not see that up front.

[the remark this answers is not in the document]
Comment (timobrembeck) [reply pyproject.toml]: Sorry for the misunderstanding, I didn't mean to remove the pinned extra, just wanted to add the non-pinned one :see_no_evil: 
Could you re-add the pinned one again? :grin: 

So our normal workflow is this:
1. Add a new dependency to the non-pinned extras
2. Run `./tools/update_dependencies.sh` to pin the version from the new dependency and calculate all sub-dependencies incl. their pinned versions

For ruff, we can do step 2. without the script by just adding `"ruff==0.3.0",` to the pinned extras manually.

[the remark this answers is not in the document]
Comment (cclauss) [reply pyproject.toml:291]: Done.

File: .circleci/config.yml [hunk 5]
```diff
@@ -85,6 +85,19 @@ jobs:
...
           command: ./tools/pylint.sh
+  ruff:
+    docker:
+      - image: cimg/python:3.11.7
+    resource_class: small
+    steps:
+      - checkout
+      - attach_workspace:
+          at: .
+      - run:
```
Comment (PeterNerlich) [inline .circleci/config.yml:96]: You probably want to also ensure the venv where ruff was installed to in *pip-install* is activated so the command can be found
```suggestion
      - run:
          name: Enable virtual environment
          command: echo "source .venv/bin/activate" >> $BASH_ENV
      - run:
```

Comment (cclauss) [reply .circleci/config.yml:96]: Ruff is written is Rust, not Python so it does not need the venv activated.

**Review by PeterNerlich — APPROVED** [review]

**Review by timobrembeck — APPROVED** [review]
I'm sorry it took me so long to get back to this! :innocent: 
Thanks a lot for your changes! :pray:
I have restarted the checks now and they're all succeeding, so I think this is good to go! :tada:

**Pull request merged**